### ISSUE TITLE
RAIL instruction

### ISSUE THREAD

**Issue by rolandschulz** [microsoft/wslg#470]
I'm interested to test RAIL for remote Linux GUI apps. Do you have any build and run instructions you could provide? It is unclear to me what is all needed to get it running and how it needs to be configured.

The only information about RAIL I could find is in this project README and https://xdc2020.x.org/event/9/contributions/611/attachments/702/1298/XDC2020_-_X11_and_Wayland_applications_in_WSL.pdf . That's why I'm asking here even though remote-Linux isn't a wslg feature.

**Comment by rolandschulz** [comment]
Yes I would like to remote over network. The README states:

> We've added support for both RAIL and VAIL to the libweston RDP backend, although for WSLg only the VAIL support is effectively used. While building WSLg, we first implemented RAIL while the necessary pieces enabling the switch to VAIL were being developed in parallel. We decided to keep that support in as it could reuse in other interesting scenarios outside of WSLg, for example for remoting application from a Pi running Linux. 

Both this and the presentation make it sound like the code required to use RAIL for remote over network is available.

**Comment by hideyukn88** [comment]
@rolandschulz, yes, this is possible, but it's out of WSLg's goal. But if you are interested in, you can build our fork version of weston on Linux (any, but internally we picked Ubuntu 20.04 on Raspberry Pi 4) and remotely connect to it from Windows using mstsc.exe on Windows 11, then you can remote Linux application running on Raspberry Pi to Windows's desktop. No code change is needed, just build, install, and run it with proper configuration. We can come up with some Wiki on this sometime in future. Thanks!

**Comment by hello-smile6** [comment]
@rolandschulz Your link is to a website with a bad certificate, could you fix it?
https://lpc.events/event/9/contributions/611/attachments/702/1298/XDC2020_-_X11_and_Wayland_applications_in_WSL.pdf

**Comment by rolandschulz** [comment]
No. I don't own that site. I just linked to it.

**Comment by hello-smile6** [comment]
@rolandschulz Could you change the link to the new domain?

**Comment by interj** [comment]
So, while it's possible to get RAIL working over network after compiling freerdp+weston, starting weston with rdp backend and connecting with mstsc.exe, there are still 2 big issues with that approach.
First one is no hardware acceleration for the remoted DISPLAY - weston with rdp backend will utilize only software rasterizer by default. From what I understand WSLg uses GfxRedir RDP channel that is a part of freerdp, but that sounds like something that would work only in VAIL scenario.
Second issue is that freerdp in RAIL mode utilizes no compression whatsoever - at least I was unable to make rfx or nscodec work, even after fiddling with winpr options for freerdp. Running 300x300px `glxgears` on 60 fps RAIL consumes around 200 Mbps of bandwidth, so it looks like these are uncompressed bitmaps being sent. With enough bandwidth applications ran very smoothly, but it was heavily limited by resolution. Surface damage detection worked well at least, static windows weren't causing almost any network movement.
I didn't try out the pulseaudio sink.

**Comment by hideyukn88** [comment]
@interj, yes, your understandings are correct. First, RDP-Backend only utilize software rendering via pixman. There was an attempt to enable GL rendering at upstream (https://gitlab.freedesktop.org/wayland/weston/-/merge_requests/165), but it never completed. But since WSLg is per-window remoting, not compositing desktop by Wayland compositor, the acceleration by GL shouldn't matter that much, but it will still incur copy by CPU. And GfxRedir protocol depends on underlying shared memory between Linux and Windows, thus, this is not possible over network (across different physical machines). Regarding to the compression, currently there is some incompatibly in compression codec with mstsc/msrdc.exe, thus we temporary disable it (https://github.com/microsoft/weston-mirror/blob/b91e7ff432af602145d216b17d3320b3345f98c5/libweston/backend-rdp/rdp.c#L1869). We can look into this when we have extra cycle, but given this is not main scenario with WSLg, it doesn't have the priority. Thanks for sharing your experience!

**Comment by interj** [comment]
Hi @hideyukn88, thanks for a quick reply.

> Regarding to the compression, currently there is some incompatibly in compression codec between FreeRDP's implementation and mstsc/msrdc.exe, thus we temporary disable it (https://github.com/microsoft/weston-mirror/blob/b91e7ff432af602145d216b17d3320b3345f98c5/libweston/backend-rdp/rdp.c#L1869). 

So as far as I see, this change disables rfx while nscodec should still be available, but judging by the bandwidth used this is not the case for me. Any place in weston or freerdp I could modify to get it to work?

**Comment by hideyukn88** [comment]
@interj, unfortunately I haven't yet figured out exact change required, and mostly the modification will be in FreeRDP side. Thanks!
